### PR TITLE
syncer/: add async flush checkpoint feature

### PR DESCRIPTION
**Pull request by lichunzhu**
cherry-pick of https://github.com/pingcap/dm/pull/595
If this PR is merged and operate stably on master branch we will then merge #595.

---

### What problem does this PR solve? 
Async flush checkpoint to improve effeciency.

### What is changed and how it works?
worker goroutine: concurrent worker count
1. The worker successfully executes a batch of jobs
2. The worker updates his checkpoint to his latest successful job pos
(The checkpoint saved by the worker goroutine is only global checkpoint, and the table checkpoint is refreshed by ddl job)

flush goroutine:
for loop
     1. Get flush type from channel. Check whether we need to update global checkpoint this time. If so, update global checkpoint with min worker pos.
     2. Flush checkpoint.

add job:
No longer scheduled wait dml job execution is complete
     a. The ddl job is executed, the wait job is completed, the checkpoint is updated, and submitted to the flush goroutine
     b. If there is no flush checkpoint for more than 30s, submit a request to the flush goroutine

fake checkpoint:
We will use added dml job location as fake checkpoint to avoid unbalanced dmls to cause an extremely small checkpoint.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported variable/fields change

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@lichunzhu please resolve conflicts

File: syncer/syncer.go [hunk 1]
```diff
@@ -974,6 +1015,10 @@ func (s *Syncer) sync(tctx *tcontext.Context, queueBucket string, db *DBConn, jo
 			if !ok {
 				return
 			}
+			if sqlJob.tp == xid {
+				lastAddedXidPos.save(sqlJob.location.Clone(), nil)
```
Comment (csuzhangxc) [inline syncer/syncer.go:1019]: Is `currentLocation` more accurate? (although `location` and `currentLocation` are the same for XID job).

Comment (lichunzhu) [reply syncer/syncer.go]: addressed in https://github.com/pingcap/dm/pull/605/commits/c1bef9fa9626eac62ce0509b7a558c2fbc379f97

File: syncer/syncer.go [hunk 2]
```diff
@@ -721,7 +745,11 @@ func (s *Syncer) addJob(job *job) error {
 		}
 		s.jobWg.Wait()
 		finishedJobsTotal.WithLabelValues("flush", s.cfg.Name, adminQueueName).Inc()
-		return s.flushCheckPoints()
```
[the remark this answers is not in the document]
Comment (lichunzhu) [reply syncer/syncer.go:723]: Newly added structure `flushHelper` may make flush checkpoint job wait.

**Comment by lichunzhu** [comment]
Won't merge

**Pull request closed without merging**